test/common/http/filter/ext_authz_test.cc

```diff
@@ -0,0 +1,264 @@
+#include <vector>
+
+#include "envoy/config/filter/http/ext_authz/v2/ext_authz.pb.validate.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/empty_string.h"
+#include "common/http/filter/ext_authz.h"
+#include "common/http/headers.h"
+#include "common/json/json_loader.h"
+#include "common/network/address_impl.h"
+#include "common/protobuf/utility.h"
+
+#include "test/mocks/ext_authz/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/local_info/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/mocks/runtime/mocks.h"
+#include "test/mocks/tracing/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/printers.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::InSequence;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::Return;
+using testing::ReturnRef;
+using testing::WithArgs;
+using testing::_;
+
+namespace Envoy {
+namespace Http {
+namespace ExtAuthz {
+
+class HttpExtAuthzFilterTest : public testing::Test {
+public:
+  HttpExtAuthzFilterTest() {}
+
+  void SetUpTest(const std::string json) {
+    envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config{};
+    MessageUtil::loadFromJson(json, proto_config);
+    config_.reset(new FilterConfig(proto_config, local_info_, stats_store_, runtime_, cm_));
+
+    client_ = new Envoy::ExtAuthz::MockClient();
+    filter_.reset(new Filter(config_, Envoy::ExtAuthz::ClientPtr{client_}));
+    filter_->setDecoderFilterCallbacks(filter_callbacks_);
+    addr_ = std::make_shared<Network::Address::Ipv4Instance>("1.2.3.4", 1111);
+  }
+
+  const std::string filter_config_ = R"EOF(
+    {
+      "grpc_service": {
+          "envoy_grpc": { "cluster_name": "ext_authz_server" }
+      },
+      "failure_mode_allow": true
+    }
+  )EOF";
+
+  FilterConfigSharedPtr config_;
+  Envoy::ExtAuthz::MockClient* client_;
+  std::unique_ptr<Filter> filter_;
+  NiceMock<MockStreamDecoderFilterCallbacks> filter_callbacks_;
+  Envoy::ExtAuthz::RequestCallbacks* request_callbacks_{};
+  TestHeaderMapImpl request_headers_;
+  Buffer::OwnedImpl data_;
+  Stats::IsolatedStoreImpl stats_store_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<Upstream::MockClusterManager> cm_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  Network::Address::InstanceConstSharedPtr addr_;
+  NiceMock<Envoy::Network::MockConnection> connection_;
+};
+
+TEST_F(HttpExtAuthzFilterTest, BadConfig) {
+  const std::string filter_config = R"EOF(
+  {
+    "failure_mode_allow": true,
+    "grpc_service": {}
+  }
+  )EOF";
+
+  envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config{};
+  MessageUtil::loadFromJson(filter_config, proto_config);
+
+  EXPECT_THROW(
+      MessageUtil::downcastAndValidate<const envoy::config::filter::http::ext_authz::v2::ExtAuthz&>(
+          proto_config),
+      ProtoValidationException);
+}
+
+TEST_F(HttpExtAuthzFilterTest, NoRoute) {
```

Comment:
For all these failure-case tests, can you parameterize them and test for both values of failure_mode_allow?

Reply:
Done. First time using parameter based testing. good to learn!